lib/db_connection.ex

```diff
@@ -817,48 +843,216 @@ defmodule DBConnection do
         :erlang.raise(kind, reason, stack)
     else
       result ->
-        conclude(conn, result)
+        {:ok, result}
     end
   end
-  def transaction(conn, fun, opts) do
-    case run(conn, &run_transaction/5, fun, meter(opts), opts) do
-      {query, {:ok, res, {kind, reason, stack}, meter}} ->
-        log(meter, :transaction, query, nil, {:ok, res})
-        :erlang.raise(kind, reason, stack)
-      {query, {:ok, res, return, meter}} ->
-        log(meter, :transaction, query, nil, {:ok, res})
-        return
-      {query, {:error, err, meter}} when query != nil ->
-        log(meter, :transaction, query, nil, {:error, err})
+  def transaction(%DBConnection{} = conn, fun, opts) do
+    case begin(conn, &run/4, opts) do
+      {:ok, conn, _} ->
+        run_transaction(conn, fun, &run/4, opts)
+      {:error, %DBConnection.TransactionError{}} ->
+        {:error, :rollback}
+      {:error, err} ->
+        raise err
+    end
+  end
+  def transaction(pool, fun, opts) do
+    case begin(pool, &checkout/4, opts) do
+      {:ok, conn, _} ->
+        run_transaction(conn, fun, &checkin/4, opts)
+      {:error, %DBConnection.TransactionError{}} ->
+        {:error, :rollback}
+      {:error, err} ->
         raise err
-      {query, return} ->
-        log(return, :transaction, query, nil)
     end
   end
 
   @doc """
-  Rollback a transaction, does not return.
+  Acquire a lock on a connection and begin a database transaction.
+
+  Return `{:ok, conn, result}` on success or `{:error, exception}` if there was
+  an error. If a new lock was acquired and there is an error, it is released.
+
+  The callback implementation should determine (using transaction status of the
+  database) the state of a transaction. If a transaction is already started,
+  according to the callback implementation, then an error tuple with a
+  `DBConnection.TransactionError` is returned.
 
-  Aborts the current transaction fun. If inside `transaction/3` bubbles
-  up to the top level.
+  This function will return a error tuple with a `DBConnection.ConnectionError`
+  when called inside a deprecated `transaction/3`.
+
+  ### Options
+
+  See module documentation. The pool and connection module may support other
+  options. All options are passed to `handle_begin/2`.
+
+  See `commit/2` and `rollback/2`.
 
   ### Example
 
-      {:error, :bar} = DBConnection.transaction(conn, fn(conn) ->
-        DBConnection.rollback(conn, :bar)
-        IO.puts "never reaches here!"
+      {:ok, conn, result} = DBConnection.begin(pool)
+      try do
+        DBConnection.execute!(conn, "SELECT * FROM table", [])
+        DBConnection.commit(conn)
+      after
+        DBConnection.rollback(conn)
```

Comment:
> Should it only be called in case execute! raises?

Yes, this is simplest/safest way to handle a transaction.



Oops i missed removing `checkout_begin` after copy and paste of docs.



> But it still seems a bit wired to me

We are depending on the database's state, not our own, so the callback implementation will use information from the last request when available.